text/0000-type-declarations.md

```diff
@@ -0,0 +1,219 @@
+
+# Detailed Design
+
+## Valid Type Declarations in PHP 5.1+
+
+- Class or interface name; e.g., `WP_REST_Request`
+- `self`, which references own class or interface name
+- `parent`, which references parent class or interface name
+- `array`, to require an array
+
+```php
+protected function foo( Bar $bar, array $items ) {
+	// $bar   must be an instance of the Bar class.
+	// $items must be passed as an array.
+}
+```
+
+## Modern Versions of PHP
+
+- The [`callable` Type Declaration](http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration) became available in PHP 5.4.
+- [Scalar Type Declarations](http://php.net/manual/en/migration70.new-features.php#migration70.new-features.scalar-type-declarations): `string`, `int`, `float`, and `bool` became available in PHP 7.0, along with support for [Return Type Declarations](http://php.net/manual/en/migration70.new-features.php#migration70.new-features.return-type-declarations) and [Strict Typing](http://php.net/manual/en/functions.arguments.php#functions.arguments.type-declaration.strict).
+- The [`iterable` Type Declaration](http://php.net/manual/en/migration71.new-features.php#migration71.new-features.iterable-pseudo-type), [Nullable Types](http://php.net/manual/en/migration71.new-features.php#migration71.new-features.nullable-types), and [`void` Return Type Declaration](http://php.net/manual/en/migration71.new-features.php#migration71.new-features.void-functions) became available in PHP 7.1.
+- The [`object` Type Declaration](http://php.net/manual/en/migration72.new-features.php#migration72.new-features.object-type) became available in PHP 7.2.
+
+## Formatting
+
+_**Note:** All of these formatting rules follow [PSR-12](https://github.com/php-fig/fig-standards/blob/master/proposed/extended-coding-style-guide.md#45-method-and-function-arguments), with one additional clarification: There MUST be one space before and after a Type Declaration; e.g., `( array $items`, not `(array$items`._
+
+### Whitespace
+
+**The additional PSR-12 clarification:**  
+There MUST be one space before and after a Type Declaration.
+
+```php
+protected function foo( Bar $bar, array $items ) {
+	// ...
+}
+```
+
+**Nullable Types:**  
+There MUST NOT be a space between the question mark and the Type Declaration.
+
+```php
+protected function foo( ?Bar $bar, ?array $items ) {
+	// ...
+}
+```
+
+#### Return Type Declarations
+
+The colon and Type Declaration MUST be on the same line as a function's closing parentheses. There MUST NOT be spaces between the function's closing parentheses and colon. There MUST be one space after the colon, followed by the Type Declaration. There MUST be one space after the Type Declaration, before the function's opening curly bracket.
```

Comment:
The "no space between closing parentheses and colon" matches PSR-12. However, having a space in there might be closer to the "use extensive spacing" approach of WordPress.

Reply:
WP coding standards is ambivalent about colons:
* for alternative control structures, a space is required before, (space or) new line after `for () :`
* ternaries require spacing on both sides of the colon
* for `case $a:` and `default`, the colon should not have a space before it

All together, the "no space between the parentheses and colon" for return types, seems to balance things out ;-)
(quite apart from making the divide between WP and the wider PHP a little smaller)